Popular_Models_Guide/Llava1.5/model_repository/llava-1.5/1/model.py

```diff
@@ -0,0 +1,206 @@
+
+import base64
+import os
+from io import BytesIO
+
+import numpy as np
+import requests as rq
+import torch
+import triton_python_backend_utils as pb_utils
+from PIL import Image
+from transformers import AutoProcessor, AutoTokenizer
+
+
+class TritonPythonModel:
+    def initialize(self, args):
+        HF_LOCATION = os.getenv("HF_LOCATION", pb_utils.get_model_dir())
+        self.image_processor = AutoProcessor.from_pretrained(HF_LOCATION)
+        self.logger = pb_utils.Logger
+        self.tokenizer = AutoTokenizer.from_pretrained(HF_LOCATION)
+        self.vocab_size = 32064
+        self.max_input_len = 2048
+
+    def _tokenize(self, prompt, num_visual_tokens):
+        chunks = prompt.split("<image>")
+        assert len(chunks) == 2, "Only support exactly one image per prompt"
+
+        return (
+            self.tokenizer.encode(chunks[0])
+            + list(range(self.vocab_size, self.vocab_size + num_visual_tokens))
+            + self.tokenizer.encode(chunks[1])[self.tokenizer.add_bos_token :]
+        )
+
+    def _parse_input(self, request, input_name, default=None):
+        input = pb_utils.get_input_tensor_by_name(request, input_name)
+        if input is not None:
+            return input.as_numpy()[0]
+
+        return default
+
+    def execute(self, requests):
+        """
+        This function receives a list of requests (`pb_utils.InferenceRequest`),
+        performs inference on every request and appends it to responses.
+        """
+        responses = []
+        for request in requests:
+            # Get INPUT0
+            image = (
+                pb_utils.get_input_tensor_by_name(request, "image")
+                .as_numpy()
+                .flatten()
+                .tolist()
+            )
+            if isinstance(image[0], bytes):
+                image = image[0].decode("utf-8")
+            pil_image = Image.open(rq.get(image, stream=True).raw).convert("RGB")
+            # Get INPUT1
+            prompt = pb_utils.get_input_tensor_by_name(request, "prompt").as_numpy()[0]
+            if isinstance(prompt, bytes):
+                prompt = prompt.decode("utf-8")
+
+            image = self.image_processor(
+                text=prompt, images=pil_image, return_tensors="np"
+            )["pixel_values"].astype(np.float16)
+            # Create inference request object
+            infer_request = pb_utils.InferenceRequest(
+                model_name="vision_encoder",
+                requested_output_names=["features"],
+                inputs=[pb_utils.Tensor("image", image)],
+            )
+
+            # Perform synchronous blocking inference request
+            vision_response = infer_request.exec()
+            response_sender = request.get_response_sender()
+
+            image_features = pb_utils.get_output_tensor_by_name(
+                vision_response, "features"
+            )
+            image_features = torch.from_dlpack(image_features.as_numpy())
```

Comment:
Can break the code block into "vison_step" and "language_step" functions to summarize the BLS procedure.

Reply:
updated